includes/admin/class-amp-cache-utilities.php

```diff
@@ -0,0 +1,251 @@
+<?php
+abstract class AMP_Cache_Utilities {
+
+	static $amp_valid_content_types = array( 'c', 'i', 'r' );
+	static $amp_cache_url_base = 'https://cdn.ampproject.org';
+	static $amp_cache_update_url_base = 'https://cdn.ampproject.org/update-ping';
+
+	/**
+	 * Hooks the 'post_updated' and 'before_delete_post' actions to determine if we
+	 * need to update the AMP Cache when a post is updated or about to be deleted.
+	 */
+	public static function amp_add_cache_update_actions() {
+		// Hooking this to the post_updated action so this will fire any time a post
+		// is updated in any way (including status transitions).
+		add_action( 'post_updated',  array( 'AMP_Cache_Utilities', 'post_updated' ), 10, 3 );
+		
+		// Hooking the call to update the cache *before* the post is updated in case we need
+		// to access any metadata for future functionality.
+		add_action( 'before_delete_post',  array( 'AMP_Cache_Utilities', 'do_amp_update_ping' ) );
+	}
+
+	/**
+	 * Called from the 'post_updated' hook when a post is updated. This hook is also fired when
+	 * a post's status changes. If the post is currently published or was not published and now is, 
+	 * then we need to update the cache.
+	 * 
+	 * This function will return true if the cache was updated successfully. It returns
+	 * false if the cache did not need to be updated *or* if the update failed for any
+	 * reason.
+	 * 
+	 * @param  int/WP_Post 	$post_id     The post that was being updated
+	 * @param  WP_Post 		$post_after  A copy of the post after the update.
+	 * @param  WP_Post		$post_before A copy of the post before the update.
+	 * @return bool         true is the cache was updated, false otherwise.
+	 */
+	public static function post_updated( $post_id, $post_after, $post_before ) {
+		// if post_status is 'publish' or was 'publish' but now is not, update cache
+		if ( ( 'publish' == $post_after->post_status ) ||
+			( ( 'publish' != $post_after->post_status ) && ( 'publish' == $post_before->post_status ) ) ) {
+			return self::do_amp_update_ping( $post_id );
+		}
+		return false;
+	}
+
+	/**
+	 * Given a post_id or WP_Post object, this function will calculate the AMP
+	 * cache update URL and ping it.
+	 * 
+	 * This function will return true if the cache was updated successfully. It returns
+	 * false if the cache did not need to be updated *or* if the update failed for any
+	 * reason.
+	 * 
+	 * @param  int/WP_Post	$post 	The post to update in the AMP cache.
+	 * @return bool       			true if the cache was updated, false otherwise.
+	 */
+	public static function do_amp_update_ping( $post ) {
+		$update_ping_url = self::get_amp_cache_update_url_for_post( $post );
+		if ( ! $update_ping_url ) {
+			return false;
+		}
+		$response = wp_remote_get( $update_ping_url );
```

Comment:
Yeah, we don't really need to wait for the response. Probably a good to just do both, i.e. make it non-blocking AND set a low timeout (since not all transports support non-blocking requests).